### PR TITLE
Avoid the fusion (in reduction) when operator data-types mismatch

### PR DESCRIPTION
**Pull request by lakhinderwalia**
Fusion of two (parallel) reductions: with mismatched data types gives a compiler error for casting of array types. This fusion of mixed data-types is now disabled.

### CODE REVIEW

File: src/targets/gpu/prepare_reduce.cpp [hunk 1]
```diff
@@ -92,11 +92,13 @@ void fuse_reductions(module& m)
     auto rs = find_parallel_reduce(find_reduce(m));
     if(rs.size() < 2)
         return;
-    // Only handle the same reduction operator for now
-    if(std::any_of(std::next(rs.begin()), rs.end(), [&](auto r) {
-           return rs.front()->name() != r->name();
+    // Only handle the same reduction operator (and its data-type) for now
+    if(std::any_of(std::next(rs.cbegin()), rs.cend(), [&](auto r) {
+           return rs.cbegin()->name() != r->name() or
```
Comment (pfultz2) [inline src/targets/gpu/prepare_reduce.cpp]: This doesn't compile, you need to use `rs.front()`.

**Comment by pfultz2** [comment]
A verify test should be added for this as well.

File: src/targets/gpu/prepare_reduce.cpp (hunk 1 above)
Comment (lakhinderwalia) [reply src/targets/gpu/prepare_reduce.cpp]: Thanks. This should fix the compile error, and logically apply the const-ness to lhs and rhs: `return (*rs.cbegin())->name() != r->name() or`  not just to rhs.

**Comment by lakhinderwalia** [comment]
> A verify test should be added for this as well.

Are we doing functionally something different here, from when we don't fuse two reduction operators? What exactly are you trying to verify that MigraphX doesn't do already!

**Comment by CharlieL7** [comment]
> > A verify test should be added for this as well.
> 
> Are we doing functionally something different here, from when we don't fuse two reduction operators? What exactly are you trying to verify that MigraphX doesn't do already!

Meaning we should have a negative test on this case. That reductions with the same name but different type are not  fused. Would be good to have the test so we don't hit the same problem in the future if things are changed.

**Comment by lakhinderwalia** [comment]
> > > A verify test should be added for this as well.
> > 
> > 
> > Are we doing functionally something different here, from when we don't fuse two reduction operators? What exactly are you trying to verify that MigraphX doesn't do already!
> 
> Meaning we should have a negative test on this case. That reductions with the same name but different type are not fused. Would be good to have the test so we don't hit the same problem in the future if things are changed.

I can add a simple verify test (a non-negative test, however).

**Review by pfultz2 — APPROVED** [review]

**Review by CharlieL7 — APPROVED** [review]
